torch/csrc/utils/python_arg_parser.h

```diff
@@ -272,6 +278,43 @@ inline const THPLayout& PythonArgs::layout(int i) {
   return *reinterpret_cast<THPLayout*>(args[i]);
 }
 
+inline Device PythonArgs::device(int i) {
+  if (!args[i]) return Device(DeviceType::CPU, -1, true);  // TODO: use CUDA if default type is a cuda type.
+  if (THPDevice_Check(args[i])) {
+    auto device = reinterpret_cast<THPDevice*>(args[i]);
+    return Device(device->device_type, device->device_index, device->is_default);
+  }
+  if (THPUtils_checkLong(args[i])) {
+    auto index = THPUtils_unpackLong(args[i]);
+    return Device(DeviceType::CUDA, index, index == -1);
+  }
+  std::string device_str = THPUtils_unpackString(args[i]);
+  std::string cpu_prefix("cpu:");
+  std::string cuda_prefix("cuda:");
```

Comment:
can those be made static? They will (each!) need an allocation every time this function is used

Reply:
Yes, good idea.